### PR TITLE
SVG Based Bubble Display Component

### LINKED ISSUE #29: Single Bubble
Component for rendering a single bubble SVG partial. Will handle events on the bubble, used by the Bubble Display component.

[Technical Document](https://iiif-timeliner.netlify.com/docs/docs-user-stories-bubble)

### LINKED ISSUE #31: Bubble Display
Wraps #29 and #30 and represents the whole bubble section of the UI. 

[Technical Document](https://iiif-timeliner.netlify.com/docs/docs-user-stories-bubble)

### PR DESCRIPTION
**Pull request by adam-digirati**
closes #31 and closes #29 

edit: automation

### CODE REVIEW

File: src/components/SingleBubble/SingleBubble.mdx [hunk 1]
```diff
@@ -3,10 +3,39 @@ title: Single bubble
...
+        style={{
+          fill: '#f6f6f6', 
+          strokeWidth: 0,
+          stroke:'rgb(0,0,0,0)'
+        }}
+      />
+      <SingleBubble 
+        x={0} 
+        width={50} 
+        height={30}
```
Comment (stephenwf) [inline src/components/SingleBubble/SingleBubble.mdx]: Demo bubble is quite small: https://deploy-preview-56--iiif-timeliner.netlify.com/docs/single-bubble-single-bubble

File: src/components/BubbleDisplay/BubbleDisplay.js [hunk 2]
```diff
@@ -23,7 +23,55 @@ class BubbleDisplay extends Component {
...
+    const projectionFactor = realWidth / maxWidth;
+    const viewBox = [x * projectionFactor, 0, width, height].join(' ');
+    const bubbles = points.map(point => ({
+      x: point.from * projectionFactor,
+      width: (point.to - point.from) * projectionFactor,
+      colour: point.colour,
+      height: Math.pow(2 / 3, point.depth - 1) * height,
+      label: point.label,
+      point: point,
+    }));
```
Comment (stephenwf) [inline src/components/BubbleDisplay/BubbleDisplay.js]: We could move the generation of `realWidth`, `maxDepth`, `maxWidth`, `projectionFactor`, `viewBox` and `bubbles` into `componentWillMount` and `componentWillReceiveProps` later on, and only re-render when the data changes. Might be needed later on if we're doing a bunch of updates to this (dragging, panning, animating etc.) No need to do it just now though, something for later

File: src/components/SingleBubble/SingleBubble.js [hunk 3]
```diff
@@ -3,21 +3,22 @@ import PropTypes from 'prop-types';
 import './SingleBubble.scss';
 
 class SingleBubble extends Component {
+  // NOTE: Changed the specification, the component doesn't really requires the y co-ordinate
```
Comment (stephenwf) [inline src/components/SingleBubble/SingleBubble.js]: Cool, comment can be removed

File: src/components/SingleBubble/SingleBubble.js [hunk 4]
```diff
@@ -28,8 +29,35 @@ class SingleBubble extends Component {
...
+      <g
+        onClick={this.onBubbleClick}
+        style={{
+          cursor: onClick ? 'pointer' : 'none',
+        }}
+      >
+        <path d={d} fill={colour} />
+        <text
+          textAnchor="middle"
+          fill="white"
```
Comment (stephenwf) [inline src/components/SingleBubble/SingleBubble.js]: Maybe something we could add to the props, with a default of white

File: src/components/BubbleDisplay/BubbleDisplay.js [hunk 5]
```diff
@@ -5,7 +5,7 @@ import './BubbleDisplay.scss';
 class BubbleDisplay extends Component {
   static propTypes = {
     /** Map of points @todo custom validator */
-    points: PropTypes.object.isRequired,
+    points: PropTypes.array.isRequired,
```
Comment (stephenwf) [inline src/components/BubbleDisplay/BubbleDisplay.js]: I think we will run into issues with using an array when it comes to updating, will be a lot of array filtering. `Object.values()` can transform the result into the array quite efficiently too, unless I'm missing another limitation

File: src/components/SingleBubble/SingleBubble.mdx (hunk 1 above)
Comment (adam-digirati) [reply src/components/SingleBubble/SingleBubble.mdx]: I add a bigger one, and a slider to modify the x, height, width...

File: src/components/SingleBubble/SingleBubble.js (hunk 3 above)
Comment (adam-digirati) [reply src/components/SingleBubble/SingleBubble.js]: see: https://github.com/digirati-co-uk/timeliner/pull/56/commits/be1f0005f9fec3e46f3c412d813075eb111a7edc

File: src/components/SingleBubble/SingleBubble.js (hunk 4 above)
Comment (adam-digirati) [reply src/components/SingleBubble/SingleBubble.js]: see https://github.com/digirati-co-uk/timeliner/pull/56/commits/d15eeb4407d6f0a9d082f69333b7536aa2aab94a

File: src/components/BubbleDisplay/BubbleDisplay.js (hunk 5 above)
Comment (adam-digirati) [reply src/components/BubbleDisplay/BubbleDisplay.js]: see https://github.com/digirati-co-uk/timeliner/pull/56/commits/c5244ff3526eee308784ea5a44890f2c8a04d5e3

**Pull request merged**